### PR TITLE
Improvements to flutter doctor JDK search

### LINKED ISSUE #8508: flutter tool wants a JDK on MacOS
## Steps to Reproduce
1. Start with a brand-new install on macOS 10.12.3 (Sierra)
1. Run `flutter doctor` for the first time, no problem.
1. Run `flutter create gallery`, get prompt:
   "To use the "java" command-line tool you need to install a JDK"
    Looks like the Gallery was created. Can't tell if flutter run works because, Santa.

## Logs
Run your application with `flutter run` and attach all the log output.
Can't do this because, Santa.

Run `flutter analyze` and attach any output of that command also.
Analyzing lib/main.dart...
No analyzer warnings! (ran in 7.2s)

## Flutter Doctor

Paste the output of running `flutter doctor` here.
[✓] Flutter (on Mac OS, channel master)
    • Flutter at /Users/eastrue/flutter
    • Framework revision 6b97757b6d (5 days ago), 2017-02-24 17:47:29
    • Engine revision 335daf1930
    • Tools Dart version 1.23.0-dev.0.0

[-] Android toolchain - develop for Android devices (Android SDK 25.0.2)
    • Android SDK at /Users/eastrue/Library/Android/sdk
    • Platform android-25, build-tools 25.0.2
    ✗ No Java Development Kit (JDK) found; you can download the JDK from https://www.oracle.com/technetwork/java/javase/downloads/.

[-] iOS toolchain - develop for iOS devices (Xcode 8.2.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 8.2.1, Build version 8C1002
    ✗ ideviceinstaller not available; this is used to discover connected iOS devices.
      Install via 'brew install ideviceinstaller'.
    ✗ ios-deploy not installed: 1.9.0 is required.
      Install via 'brew install ios-deploy'.

[✓] Android Studio (version 2.2)
    • Android Studio at /Users/eastrue/Applications/Android Studio.app/Contents
    • Gradle version 2.14.1

[✓] Connected devices
    • iPhone SE • D6E31428-32E2-4AEF-8C87-7719F568D94A • ios • iOS 10.2 (simulator)

> For more information about diagnosing and reporting Flutter bugs, please see [https://flutter.io/bug-reports/](https://flutter.io/bug-reports/).

### LINKED ISSUE #8521: flutter doctor doesn't recognize JDK in PATH
## Steps to Reproduce

1. Install Android Studio, with its bundled JDK.
1. Run `export PATH=/Applications/Android\ Studio.app/Content/jre/jdk/Contents/Home/bin:$PATH`
1. Also `vi ~/.bash_profile` to add this and `source`.
1. With Android Studio running, run `flutter doctor`.
Output shows 
```
[-] Android toolchain - develop for Android devices (Android SDK 25.0.2)
    • Android SDK at /Users/eastrue/Library/Android/sdk
    • Platform android-25, build-tools 25.0.2
    ✗ No Java Development Kit (JDK) found; you can download the JDK from https://www.oracle.com/technetwork/java/javase/downloads/.
```

It also shows no connected device, although there is one and it has shown up in the output today.

## Logs
n/a, but `echo $PATH` yields:
`/Applications/Android Studio.app/Content/jre/jdk/Contents/Home/bin:/Applications/AndroidStudio.app/Content/jre/jdk/Contents/Home/bin:/Users/eastrue/flutter/bin:/usr/local/git/current/bin:/usr/local/bin:/usr/bin:/bin:/usr/local/sbin:/usr/sbin:/sbin`

## Flutter Doctor

[✓] Flutter (on Mac OS, channel master)
    • Flutter at /Users/eastrue/flutter
    • Framework revision 6b97757b6d (5 days ago), 2017-02-24 17:47:29
    • Engine revision 335daf1930
    • Tools Dart version 1.23.0-dev.0.0

[-] Android toolchain - develop for Android devices (Android SDK 25.0.2)
    • Android SDK at /Users/eastrue/Library/Android/sdk
    • Platform android-25, build-tools 25.0.2
    ✗ No Java Development Kit (JDK) found; you can download the JDK from https://www.oracle.com/technetwork/java/javase/downloads/.

[✓] iOS toolchain - develop for iOS devices (Xcode 8.2.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 8.2.1, Build version 8C1002
    • ios-deploy 1.9.1

[✓] Android Studio (version 2.2)
    • Android Studio at /Users/eastrue/Applications/Android Studio.app/Contents
    • Gradle version 2.14.1

[✓] IntelliJ IDEA Community Edition (version 2016.3.1)
    • Dart plugin version 163.13137
    • Flutter plugin version 0.1.11.1

[✓] Connected devices
    • None

> For more information about diagnosing and reporting Flutter bugs, please see [https://flutter.io/bug-reports/](https://flutter.io/bug-reports/).

### PR DESCRIPTION
**Pull request by johnmccutchan**
- [x] Add custom logic on MacOS to determine if Java is installed before invoking `java`.
- [x] Check JAVA_HOME, platform specific logic, and finally PATH to locate the `java` executable.
- [x] Improved doctor messages.

Fixes #8508
Fixes #8521

@tvolkert @goderbauer

### CODE REVIEW

**Comment by sethladd** [comment]
Would this also help with #8515 ?

File: packages/flutter_tools/lib/src/android/android_workflow.dart [hunk 1]
```diff
@@ -42,63 +106,56 @@ class AndroidWorkflow extends DoctorValidator implements Workflow {
...
-      }
+      return new ValidationResult(ValidationType.partial, messages, statusInfo: sdkVersionText);
+    }
+
+    // Now check for the JDK.
+    final String javaBinary = _findJavaBinary();
+    if (javaBinary == null) {
+      messages.add(new ValidationMessage.error(
+          'No Java Development Kit (JDK) found; You must have the environment '
+          'variable JAVA_HOME set and the java binary in your PATH. '
```
Comment (goderbauer) [inline packages/flutter_tools/lib/src/android/android_workflow.dart]: Do you really need both? I have been running under Windows so far without JAVA_HOME set to anything (this is how the default JDK installer sets up Java on Windows).

**Comment by Hixie** [comment]
Fixed the TODO that I'd left, and caught another error that the dependency checker might run into.

File: packages/flutter_tools/lib/src/android/android_workflow.dart (hunk 1 above)
Comment (johnmccutchan) [inline packages/flutter_tools/lib/src/android/android_workflow.dart]: @goderbauer You don't strictly need both but if a user doesn't have a working Java environment I think it's better to give them more direction here.

**Comment by johnmccutchan** [comment]
@sethladd No, that is something separate.

**Comment by johnmccutchan** [comment]
@Hixie commenting on the wrong PR?

File: packages/flutter_tools/lib/src/android/android_workflow.dart [hunk 2]
```diff
@@ -23,13 +26,74 @@ class AndroidWorkflow extends DoctorValidator implements Workflow {
...
+        final List<String> javaHomeOutputSplit = javaHomeOutput.split('\n');
+        if ((javaHomeOutputSplit != null) && (javaHomeOutputSplit.length > 0)) {
+          final String javaHome = javaHomeOutputSplit[0].trim();
+          return fs.path.join(javaHome, 'bin', 'java');
+        }
+      }
+    }
+
+    // Fallback to PATH based lookup.
+    return os.which(_kJavaExecutable)?.path;
```
Comment (tvolkert) [inline packages/flutter_tools/lib/src/android/android_workflow.dart]: Same comment about Windows

File: packages/flutter_tools/lib/src/android/android_workflow.dart [hunk 3]
```diff
@@ -42,63 +106,56 @@ class AndroidWorkflow extends DoctorValidator implements Workflow {
...
         ));
       }
-    } else {
-      type = ValidationType.partial;
-
-      messages.add(new ValidationMessage('Android SDK at ${androidSdk.directory}'));
 
-      if (androidSdk.latestVersion != null) {
-        sdkVersionText = 'Android SDK ${androidSdk.latestVersion.buildToolsVersionName}';
+      return new ValidationResult(ValidationType.missing, messages, statusInfo: sdkVersionText);
```
Comment (tvolkert) [inline packages/flutter_tools/lib/src/android/android_workflow.dart]: `sdkVersionText` is `null` at this point, so just don't pass `statusInfo`?

File: packages/flutter_tools/lib/src/android/android_workflow.dart [hunk 4]
```diff
@@ -23,13 +26,74 @@ class AndroidWorkflow extends DoctorValidator implements Workflow {
...
+  static const String _kJavaExecutable = 'java';
+  static const String _kJdkDownload = 'https://www.oracle.com/technetwork/java/javase/downloads/';
+
+  /// First sniff JAVA_HOME, then fallback to PATH.
+  String _findJavaBinary() {
+
+    final String javaHomeEnv = platform.environment[_kJavaHomeEnvironmentVariable];
+    if (javaHomeEnv != null) {
+      // Trust JAVA_HOME.
+      return fs.path.join(javaHomeEnv, 'bin', 'java');
```
Comment (tvolkert) [inline packages/flutter_tools/lib/src/android/android_workflow.dart]: @goderbauer, do we have to worry about it being java.bat or java.exe on Windows?

File: packages/flutter_tools/lib/src/android/android_workflow.dart [hunk 5]
```diff
@@ -23,13 +26,74 @@ class AndroidWorkflow extends DoctorValidator implements Workflow {
...
+
+    // MacOS specific logic to avoid popping up a dialog window.
+    // See: http://stackoverflow.com/questions/14292698/how-do-i-check-if-the-java-jdk-is-installed-on-mac.
+    if (platform.isMacOS) {
+      String javaHomeOutput;
+      try {
+        javaHomeOutput = runCheckedSync(<String>['/usr/libexec/java_home'],
+                                        hideStdout: true);
+      } catch (_) { /* ignore */ }
+      if (javaHomeOutput != null) {
```
Comment (tvolkert) [inline packages/flutter_tools/lib/src/android/android_workflow.dart]: Just do this inside the `try` block?

Comment (johnmccutchan) [inline packages/flutter_tools/lib/src/android/android_workflow.dart]: Done.

File: packages/flutter_tools/lib/src/android/android_workflow.dart (hunk 3 above)
Comment (johnmccutchan) [inline packages/flutter_tools/lib/src/android/android_workflow.dart]: Done.

File: packages/flutter_tools/lib/src/android/android_workflow.dart (hunk 4 above)
Comment (goderbauer) [inline packages/flutter_tools/lib/src/android/android_workflow.dart]: As long as this string is executed by the process manager we should be good on Windows (the process manager abstracts away the fact that an executable might have a different extensions on Windows. Looks like this string is passed to `processManager.runSync` below, so we should be good here.

**Comment by tvolkert** [comment]
LGTM

File: packages/flutter_tools/lib/src/android/android_workflow.dart (hunk 2 above)
Comment (goderbauer) [inline packages/flutter_tools/lib/src/android/android_workflow.dart]: We do have a windows implementation for `os.which`, so this should be ok.

**Comment by goderbauer** [comment]
LGTM

**Pull request merged**